pkg/cloudprovider/types.go

```diff
@@ -59,25 +59,28 @@ type NodeGroup interface {
 type Capacity interface {
 	// Create a set of nodes to fulfill the desired capacity given constraints.
 	Create(context.Context, *Constraints) ([]Packing, error)
-
-	// GetTopologyDomains returns a list of topology domains supported by the
-	// cloud provider for the given key.
-	// For example, GetTopologyDomains("zone") -> [ "us-west-2a", "us-west-2b" ]
-	// This enables the caller to to build Constraints for a known set of
-	GetTopologyDomains(context.Context, TopologyKey) ([]string, error)
+	// GetZones returns a list of zones supported by the cloud provider.
+	GetZones(context.Context) ([]string, error)
 }
 
-// Constraints lets the controller define the desired capacity,
-// avalability zone, architecture for the desired nodes.
+// Constraints for an efficient binpacking solution of pods onto nodes, given
+// overhead and node constraints.
 type Constraints struct {
-	// Pods is a list of equivalently schedulable pods to be efficiently
-	// binpacked.
+	NodeConstraints
+	// Pods is a list of equivalently schedulable pods to be binpacked.
 	Pods []*v1.Pod
-	// Overhead resources per node from system resources such a kubelet and
-	// daemonsets.
+	// Overhead resources per node from system resources such as kubelet and daemonsets.
 	Overhead v1.ResourceList
-	// Topology constrains the topology of the node, e.g. "zone".
-	Topology map[TopologyKey]string
+}
+
+// NodeConstraints constrain each node within the binpacking solution.
+type NodeConstraints struct {
+	// Labels may define additional cloud provider specific provisioning behavior.
+	Labels map[string]string
+	// Zone optionally constrains the zone where the node is launched.
+	Zone *string
```

Comment:
Can a node be constrained to multiple zones? Is zone selection supposed to have occurred before this struct is filled?

Reply:
I was thinking of this as either specify zone or allow the controller to specify the zone. There might be a usecase as you suggest, but I'm failing to see it.

Reply:
If this is the way to limit to a specific set of zones, I could see something here. Like in the case of 5-zone regions, it would be the list of the 3 zones you actually want to be in (so you don't start having to deal with a 4th or 5th zone). Or if for some reason you have picked 2 zones you're going to be in and you want to be in those two? Maybe I'm misunderstanding what this field is for though.

Reply:
I think you're right here. We to be able to limit instance type and zone at the Provisioner level. These will then need to be pass through this constraints API. This should be a slice of options. I'll sort this out in a followon.